## Step 1: Import Libraries

- `pandas` and `numpy` help us work with data.

- `matplotlib` and `seaborn` are used to create charts and graphs.

- From `scikit-learn`, we import tools to split data, build a random forest model, and check how well the model performs.

- `compute_class_weight` helps us deal with class imbalance in the data.

- `thefuzz` is used for fuzzy string matching—useful when text isn’t exactly the same.

- `re` is Python’s tool for working with text patterns using regular expressions.

In [ ]:
# imported libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score
from sklearn.utils.class_weight import compute_class_weight
from thefuzz import process
import re
import io


## References

* [Scikit-learn](https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html) Random Forest Classifier documentation  
* [Scikit-learn](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html) train-test split documentation  
* [Scikit-learn](https://scikit-learn.org/stable/modules/generated/sklearn.utils.class_weight.compute_class_weight.html) compute class weight documentation  
* [Scikit-learn](https://scikit-learn.org/stable/modules/model_evaluation.html) metrics documentation (accuracy, confusion matrix, classification report)  
* [Pandas documentation](https://pandas.pydata.org/docs/) for data manipulation  
* [NumPy documentation](https://numpy.org/doc/) for numerical computing  
* [Matplotlib](https://matplotlib.org/stable/index.html) official documentation for plotting  
* [Seaborn](https://seaborn.pydata.org/) official documentation for statistical data visualization  
* [Medium article](https://medium.com/@jaimejcheng/data-exploration-and-visualization-with-seaborn-pair-plots-40e6d3450f6d) for Seaborn pair plots  
* [TheFuzz GitHub](https://github.com/seatgeek/thefuzz) for fuzzy string matching in Python  
* [Medium article](https://towardsdatascience.com/fuzzy-string-matching-in-python-68f240d910fe) explaining fuzzy matching in Python  
* [Hands-On Machine Learning book](https://www.amazon.com/Hands-Machine-Learning-Scikit-Learn-TensorFlow/dp/1491962291) for machine learning workflows and best practices  
* [UCI Machine Learning Repository](https://archive.ics.uci.edu/ml/index.php) for accessing open datasets  




## Step 2: Load and Prepare the Dataset

- Load the dataset from the Excel file `Cervical Cancer Datasets_.xlsx`.
- Save it as a CSV file named `cervical_cancer.csv` to make it easier to work with.
- Load the CSV file into a Pandas DataFrame called `data` for further analysis.

Note: We use `index=False` when saving to prevent Pandas from adding row numbers as a column.


In [ ]:

# Load Excel and Save as CSV
excel_path = 'Cervical Cancer Datasets_.xlsx'
csv_path = 'cervical_cancer.csv'

df = pd.read_excel(excel_path)
df.to_csv(csv_path,index=False)

# Load CSV (use header=None only if the dataset has no headers)
data = pd.read_csv(csv_path)  # Remove `header=None` if headers are present

data


Patient ID  Age  Sexual Partners  First Sexual Activity Age  \
0       P0001   18                4                         15   
1       P0002   15                1                         14   
2       P0003   34                1                          9   
3       P0004   52                5                         16   
4       P0005   46                3                         21   
..        ...  ...              ...                        ...   
95      P0096   31                4                         16   
96      P0097   35                5                         11   
97      P0098   35                1                         18   
98      P0099   31                1                         20   
99      P0100   59                6                         22   

   HPV Test Result Pap Smear Result Smoking Status STDs History     Region  \
0         NEGATIVE                N              N            Y   Pumwani    
1         POSITIVE                N              Y            Y  Kakamega    
2         POSITIVE                N              N            Y   Machakos   
3         POSITIVE                N              Y            N      Embu    
4         POSITIVE                N              N            N    Mombasa   
..             ...              ...            ...          ...        ...   
95        POSITIVE                Y              Y            N   Machakos   
96        NEGATIVE                N              Y            N    Mombasa   
97        POSITIVE                N              Y            N    Kericho   
98        POSITIVE                Y              Y            N      Embu    
99        POSITIVE                Y              N            Y    Kitale    

   Insrance Covered Screening Type Last  \
0                 Y           PAP SMEAR   
1                 N             HPV DNA   
2                 N             HPV DNA   
3                 Y             HPV DNA   
4                 N             HPV DNA   
..              ...                 ...   
95                N           PAP SMEAR   
96                N           PAP SMEAR   
97                N             HPV DNA   
98                Y             HPV DNA   
99                Y             HPV DNA   

                                 Recommended Action Unnamed: 12  
0   REPEAT PAP SMEAR IN 3 YEARS AND FOR HPV VACCINE         NaN  
1              FOR HPV VACCINE AND SEXUAL EDUCATION         NaN  
2              FOR HPV VACCINE AND SEXUAL EDUCATION         NaN  
3   FOR HPV VACCINE, LIFESTYLE AND SEXUAL EDUCATION         NaN  
4              FOR HPV VACCINE AND SEXUAL EDUCATION         NaN  
..                                              ...         ...  
95                  FOR COLPOSCOPY BIOPSY, CYTOLOGY         NaN  
96                      REPEAT PAP SMEAR IN 3 YEARS         NaN  
97     FOR ANUAL FOLLOW UP AND PAP SMEAR IN 3 YEARS         NaN  
98                 FOR COLPOSCOPY BIOPSY, CYTOLOGY          NaN  
99           FOR COLPOSCOPY BIOPSY, CYTOLOGY +/-TAH         NaN  

[100 rows x 13 columns]

### Flag Duplicates

The `Patient ID` is treated as the unique identifier for each individual.

We check for duplicate records by comparing all other columns **except `Patient ID`**. If rows have the same values across at least 80% of the fields, they are flagged as potential duplicates.

A new column `is_duplicate` is added to the dataset to mark these cases.

All flagged records are then exported to `flagged_duplicates.csv` for manual checking and review, ensuring no important data is accidentally removed.


In [ ]:
# Flag duplicates and export them for manual checking
data['is_duplicate'] = data.duplicated(subset=df.columns.difference(['Patient ID']), keep=False)
duplicates_df = data[data['is_duplicate'] == True]
duplicates_df.to_csv('flagged_duplicates.csv', index=False)

print(data.columns)
print("\n")
print(f"The data set has ( Total rows: {len(data)}, Total columns: {len(data.columns)})")


### Inspect and Verify Data Quality

In this step, we perform a basic data audit:

- Load the dataset from the CSV file.
- Preview the data using `.head()` to understand its structure.
- Use `.info()` to check data types and identify columns with missing values.
- Count missing values per column to assess data completeness.
- Print all unique values for each column to inspect potential inconsistencies or unexpected entries (e.g., typos in categorical fields).
- Finally, we print the value counts for each column — including missing values — to evaluate how frequently each value appears. This helps identify:
  - Class imbalance
  - Invalid or rare values
  - Columns that may require cleaning or transformation

In [ ]:
# Load the raw data
data = pd.read_csv("cervical_cancer.csv")

# Preview the first few rows
print(data.head())

# Overview of column names, data types, and non-null counts
print(data.info())

# Check for missing values in each column
print("Missing values per column:\n", data.isnull().sum())

# Show unique values and value counts for each column
for column in data.columns:
    print(f"--- Unique values for: {column} ---")
    print(data[column].unique())
    print(f"\n--- Value Counts for: {column} ---")
    print(data[column].value_counts(dropna=False))
    print("\n")



  Patient ID  Age  Sexual Partners  First Sexual Activity Age HPV Test Result  \
0      P0001   18                4                         15        NEGATIVE   
1      P0002   15                1                         14        POSITIVE   
2      P0003   34                1                          9        POSITIVE   
3      P0004   52                5                         16        POSITIVE   
4      P0005   46                3                         21        POSITIVE   

  Pap Smear Result Smoking Status STDs History     Region Insrance Covered  \
0                N              N            Y   Pumwani                 Y   
1                N              Y            Y  Kakamega                 N   
2                N              N            Y   Machakos                N   
3                N              Y            N      Embu                 Y   
4                N              N            N    Mombasa                N   

  Screening Type Last                       

## Step 3: Exploratory Data Analysis (EDA)

Exploratory Data Analysis (EDA) helps us deeply understand the structure, quality, and relationships within the data before applying any machine learning algorithms.

### EDA Goals:
- Identify variable types (categorical, numerical, binary)
- Explore distributions of key features
- Visualize relationships and potential predictors
- Detect class imbalance in the target variable
- Spot outliers or anomalies that may impact modeling
- Support decisions for further data cleaning or transformation

---

### Feature & Data Type Overview

We begin by checking the types of each column and the number of unique values. This helps us distinguish between categorical and numerical features and spot columns that need transformation or standardization.

---

### Known Data Quality Discrepancies

During initial inspection, several inconsistencies and typos were identified:

#### Column: `HPV Test Result`
- `"POSITIVE\n"` — extra newline character
- `"NEGAGTIVE"` — misspelling of `"NEGATIVE"`

#### Column: `Region`
- Inconsistent casing: `"Mombasa"` vs `"MOMBASA"`, `"Kitale"` listed multiple times with trailing spaces
- Spacing and capitalization differences

#### Column: `Recommended Action`
- Same recommendation repeated with:
  - Extra spaces: `"REPEAT PAP SMEAR IN 3 YEARS "` vs `"REPEAT PAP SMEAR IN 3 YEARS"`
  - Misspellings: `"BIOSPY"` instead of `"BIOPSY"`, `"COLOSCOPY"` instead of `"COLPOSCOPY"`
  - Concatenation: `"FORCOLPOSCOPY, CYTOLOGY THEN LASER THERAPY"`

#### Column: `Unnamed: 12`
- Largely empty (mostly `NaN`) — candidate for removal.

#### Column: `Insrance Covered`
- Misspelled — should be renamed to `"Insurance Covered"`

---

### 1. Outcome Class Imbalance

We check whether the target variable (e.g., HPV Test Result or Pap Smear Result) is imbalanced.

> Why it's important: Imbalanced datasets can mislead the model into favoring the dominant class, leading to poor generalization.

---

### 2. Distribution of Numeric Features

We explore numerical features like `Age`, `First Sexual Activity Age`, and `Sexual Partners` to check their distribution and spot outliers using histograms and boxplots.

This step helps:
- Identify skewness
- Detect abnormal values
- Guide transformation decisions

---

### 3. Binary Features Overview

Binary/categorical columns such as `Smoking Status`, `STDs History`, and `Insurance Covered` are analyzed to ensure:
- Consistency (e.g., "YES"/"NO" vs "Y"/"N")
- No missing or ambiguous values

Visualizing their distribution gives insight into population characteristics.

---

### 4. Categorical Variable Exploration

We examine values in columns like:
- `Region` – Check patient spread across different geographical locations
- `Screening Type Last` – Evaluate most commonly used screening methods
- `Recommended Action` – Understand typical medical advice given after testing

This helps us assess frequency, diversity, and inconsistencies in responses.

---

### 5. Correlation Matrix for Numerical Features

A correlation matrix (heatmap) shows how numerical features relate to each other. This helps detect:
- Redundant or highly correlated variables
- Relationships that may influence the model

Features explored include:
- `Age`
- `First Sexual Activity Age`
- `Sexual Partners`

---

### 6. Summary of Key Issues to Clean

| Column                 | Issue Type             | Action Needed                     |
|------------------------|------------------------|-----------------------------------|
| `HPV Test Result`      | Typos, extra characters| Normalize strings                 |
| `Region`               | Casing, spacing issues | Map to consistent labels          |
| `Recommended Action`   | Misspellings, spacing  | Standardize text                  |
| `Unnamed: 12`          | Mostly empty           | Drop this column                  |
| `Insrance Covered`     | Misspelled             | Rename to `Insurance Covered`     |

---

### Bonus: Scikit-learn Data Cleaning Pipeline (Optional)

To make cleaning reusable and organized, we can define custom classes using `scikit-learn` transformers and compose them into a pipeline.

Example concept:
- Create a transformer to clean and standardize `HPV Test Result`
- Chain transformers for cleaning multiple fields

Using pipelines ensures our data transformations are:
- Reproducible
- Modular
- Easily integrated into training workflows

---

### EDA Summary

- Confirmed the presence of text inconsistencies, typos, and formatting issues in categorical columns.
- Several numerical features show outliers or unusual values.
- The target variable is imbalanced, which may require resampling.
- These insights guide the next step: comprehensive data cleaning and transformation.


###  3.1: Initial Inspection of Raw Data

In this step, we load the raw cervical cancer dataset and conduct an initial inspection to identify data quality issues that need to be addressed during cleaning. Here's what each section does:

- **Suppress Warnings**: 
  We turn off warning messages to keep the notebook output clean using `warnings.filterwarnings('ignore')`.

- **Load Data**: 
  The dataset is loaded from a CSV file (`cervical_cancer.csv`) using `pandas.read_csv()`, and we print the shape to understand how many records (rows) and features (columns) it contains.

- **Column Overview**: 
  We print all the column names using `df.columns.tolist()` to:
  - Identify any unnecessary columns (like auto-generated ones: `Unnamed: ...`)
  - Spot typos in column names

- **Sample Problematic Data**: 
  We inspect three important columns:
  - `HPV Test Result`
  - `Region`
  - `Recommended Action`
  
  This helps us identify:
  - Inconsistent values (e.g., different spellings, casing, or newline characters)
  - Redundancy or unclear labels

- **Check for Missing Values**: 
  Finally, we run `df.isnull().sum()` to count how many missing values exist in each column. This is crucial for deciding whether to impute, drop, or otherwise handle incomplete data during cleaning.

This early scan lays the foundation for designing a robust data cleaning process and informs which fields require attention for standardization, correction, or removal.


### 🔍 3.1: Initial Inspection of Raw Data

In this step, we load the raw cervical cancer dataset and perform a quick inspection to identify data quality issues before cleaning. Specifically, we:

- Suppress warning messages for a cleaner notebook display.
- Load the dataset from a CSV file and print its shape to understand how many rows and columns it contains.
- Print the column names to check for any typos or unnecessary fields.
- Display unique values from key columns like `HPV Test Result`, `Region`, and a sample of `Recommended Action` values to identify inconsistencies (e.g., typos, inconsistent formatting, or unexpected categories).
- Check for missing values in each column to determine which features may require imputation or removal during cleaning.

This preliminary scan helps us understand what kind of cleaning and standardization will be necessary to prepare the data for analysis.


In [ ]:
import pandas as pd
import numpy as np
import warnings
warnings.filterwarnings('ignore')

# Load the raw data
print("Loading raw data...")
df = pd.read_csv('cervical_cancer.csv')
print(f"Original dataset shape: {df.shape}")
print("\n" + "="*50)
print("BEFORE CLEANING - Data Quality Issues:")
print("="*50)

# Show original data quality issues
print("1. Column names:")
print(df.columns.tolist())

print("\n2. Sample of problematic data:")
print("HPV Test Result unique values:", df['HPV Test Result'].unique())
print("Region unique values:", df['Region'].unique())
print("Recommended Action samples:")
for action in df['Recommended Action'].unique()[:5]:
    print(f"  - '{action}'")

print("\n3. Missing values:")
print(df.isnull().sum())

  Patient ID  Age  Sexual Partners  First Sexual Activity Age HPV Test Result  \
0      P0001   18                4                         15        NEGATIVE   
1      P0002   15                1                         14        POSITIVE   
2      P0003   34                1                          9        POSITIVE   
3      P0004   52                5                         16        POSITIVE   
4      P0005   46                3                         21        POSITIVE   

  Pap Smear Result Smoking Status STDs History     Region Insrance Covered  \
0                N              N            Y   Pumwani                 Y   
1                N              Y            Y  Kakamega                 N   
2                N              N            Y   Machakos                N   
3                N              Y            N      Embu                 Y   
4                N              N            N    Mombasa                N   

  Screening Type Last                       

### 3.2: Data Cleaning and Standardization

In this step, we clean and standardize the cervical cancer dataset to improve data quality, remove inconsistencies, and prepare it for further analysis and modeling.

---

#### Step-by-Step Cleaning Breakdown:

1. **Create a Copy**  
   - We create a copy of the original DataFrame (`clean_df = df.copy()`) to preserve the raw data for reference or backup.

2. **Remove Unnecessary Columns**  
   - We drop columns like `Unnamed: 12`, which are often artifacts of Excel exports and contain mostly `NaN` or irrelevant data.

3. **Fix Column Name Typos**  
   - Rename incorrectly spelled column headers.  
     Example: `"Insrance Covered"` → `"Insurance Covered"`

4. **Clean `HPV Test Result` Column**  
   - Convert values to uppercase and strip whitespace.
   - Replace typos and inconsistent formats like:
     - `"NEGAGTIVE"` → `"NEGATIVE"`
     - `"POSITIVE\n"` → `"POSITIVE"`
   - This standardization allows consistent analysis and visualization.

5. **Clean and Standardize `Region` Names**  
   - Remove inconsistencies caused by case sensitivity and trailing spaces.
   - Use a mapping dictionary to unify variations:
     - `"MOMBASA"`, `"mombasa "` → `"Mombasa"`
     - `"PUMWANI"`, `"Pumwani "` → `"Pumwani"` etc.

6. **Clean Binary Columns (Y/N Values)**  
   - Columns like:
     - `Pap Smear Result`, `Smoking Status`, `STDs History`, `Insurance Covered`
   - We standardize these by:
     - Converting to uppercase
     - Stripping whitespace  
   This ensures we don’t treat `"Y"`, `"y "`, and `" Y"` as different values.

7. **Clean `Screening Type Last` Column**  
   - Normalize string values using `upper()` and `strip()` to handle inconsistent formatting.

8. **Handle Age Anomalies**  
   - Convert `Age` to numeric and clip unrealistic values:
     - Any age `< 10` or `> 100` is corrected to fall within that range.

9. **Fix First Sexual Activity Age Conflicts**  
   - Convert to numeric and check for logical errors:
     - Drop rows where `First Sexual Activity Age > Age`.

10. **Clean `Sexual Partners` Column**  
    - Convert to numeric, drop rows with non-numeric or missing values.

11. **Final String Cleanup**  
    - Strip all string-type columns of whitespace for consistency.

---

#### Summary After Cleaning:

- Original Shape: Shows the number of rows and columns before cleaning.
- Cleaned Shape: Shows the final shape after dropping or fixing records.
- Confirmed Cleaned Columns:
  - `HPV Test Result` — Cleaned and standardized
  - `Region` — Normalized region names
  - Binary and age-related features — Cleaned, standardized, and validated

---

By thoroughly cleaning and validating the data, we ensure higher accuracy in modeling, better visualization outcomes, and more meaningful insights during the next step: **Exploratory Data Analysis (EDA)**.


In [87]:
# Create a copy for cleaning
clean_df = df.copy()

print("\n" + "="*50)
print("APPLYING CLEANING TRANSFORMATIONS:")
print("="*50)


# 1. Remove unnecessary columns
print("1. Removing unnecessary columns...")
columns_to_drop = ['Unnamed: 12']
for col in columns_to_drop:
    if col in clean_df.columns:
        clean_df.drop(columns=[col], inplace=True)
        print(f"   Dropped column: {col}")
        

# 2. Fix column name typos
print("\n2. Fixing column name typos...")
column_renames = {
    'Insrance Covered': 'Insurance Covered'
}
clean_df.rename(columns=column_renames, inplace=True)
for old, new in column_renames.items():
    print(f"   Renamed '{old}' to '{new}'")
    

# 3. Clean HPV Test Result
print("\n3. Cleaning HPV Test Result...")
print(f"   Before: {clean_df['HPV Test Result'].unique()}")

clean_df['HPV Test Result'] = clean_df['HPV Test Result'].astype(str).str.upper().str.strip()



# Fix the problematic characters - using raw strings to avoid escape issues
hpv_replacements = {
    'NEGAGTIVE': 'NEGATIVE',
    'POSTIVE': 'POSITIVE',
    'NEGATVIE': 'NEGATIVE',
    'NEGATVE': 'NEGATIVE',
    'NEGATIVE\n':'NEGATIVE',
    'NEGATIVE ':'NEGATIVE',
    ' NEGATIVE ':'NEGATIVE',
    ' POSITIVE ':'POSITIVE',
    'POSITIVE\n':'POSITIVE'
}

# Handle newline characters separately
clean_df['HPV Test Result'] = clean_df['HPV Test Result'].str.replace('\n', '', regex=False)
clean_df['HPV Test Result'] = clean_df['HPV Test Result'].str.replace('\\n', '', regex=False)

# Apply other replacements
clean_df['HPV Test Result'] = clean_df['HPV Test Result'].replace(hpv_replacements)

print(f"   After: {clean_df['HPV Test Result'].unique()}")


# 4. Clean and standardize Region names
print("\n4. Cleaning and standardizing Region names...")
print(f"   Before: {sorted(clean_df['Region'].unique())}")

# First, strip whitespace and normalize case
clean_df['Region'] = clean_df['Region'].astype(str).str.strip()

# Create comprehensive region mapping to handle all variations
region_mapping = {
    # Mombasa variations
    'mombasa': 'Mombasa',
    'MOMBASA': 'Mombasa',
    'Mombasa': 'Mombasa',
    'mombasa ': 'Mombasa',
    'Mombasa ': 'Mombasa',
    'MOMBASA ': 'Mombasa',


      # Pumwani variations
    'pumwani': 'Pumwani',
    'PUMWANI': 'Pumwani',
    'Pumwani': 'Pumwani',
    'pumwani ': 'Pumwani',
    'Pumwani ': 'Pumwani',
    'PUMWANI ': 'Pumwani',
    
    # Embu variations
    'embu': 'Embu',
    'EMBU': 'Embu',
    'Embu': 'Embu',
    'embu ': 'Embu',
    'Embu ': 'Embu',
    'EMBU ': 'Embu',
    
    # Kakamega variations
    'kakamega': 'Kakamega',
    'KAKAMEGA': 'Kakamega',
    'Kakamega': 'Kakamega',
    'kakamega ': 'Kakamega',
    'Kakamega ': 'Kakamega',
    'KAKAMEGA ': 'Kakamega',
    
    # Machakos variations
    'machakos': 'Machakos',
    'MACHAKOS': 'Machakos',
    'Machakos': 'Machakos',
    'machakos ': 'Machakos',
    'Machakos ': 'Machakos',
    'MACHAKOS ': 'Machakos',
    
    # Nakuru variations
    'nakuru': 'Nakuru',
    'NAKURU': 'Nakuru',
    'Nakuru': 'Nakuru',
    'NAKURU ': 'Nakuru',
    
    # Moi variations
    'moi': 'Moi',
    'MOI': 'Moi',
    'Moi': 'Moi',
    'moi ': 'Moi',
    'Moi ': 'Moi',
    'MOI ': 'Moi',
    
  
  
    # Loitoktok variations
    'loitoktok': 'Loitoktok',
    'LOITOKTOK': 'Loitoktok',
    'Loitoktok': 'Loitoktok',
    'loitoktok ': 'Loitoktok',
    'Loitoktok ': 'Loitoktok',
    'LOITOKTOK ': 'Loitoktok',
    
    # Garissa variations
    'garissa': 'Garissa',
    'GARISSA': 'Garissa',
    'Garissa': 'Garissa',
    'garissa ': 'Garissa',
    'Garissa ': 'Garissa',
    'GARISSA ': 'Garissa',
    
    # Kericho variations
    'kericho': 'Kericho',
    'KERICHO': 'Kericho',
    'Kericho': 'Kericho',
    'kericho ': 'Kericho',
    'Kericho ': 'Kericho',
    'KERICHO ': 'Kericho',
    
    # Kitale variations
    'kitale': 'Kitale',
    'KITALE': 'Kitale',
    'Kitale': 'Kitale',
    'kitale ': 'Kitale',
    'Kitale ': 'Kitale',
    'KITALE ': 'Kitale'
}



# Apply region mapping
clean_df['Region'] = clean_df['Region'].replace(region_mapping)

print(f"   After: {sorted(clean_df['Region'].unique())}")


# 6. Clean binary columns (Y/N values)
print("\n6. Cleaning binary columns...")
binary_columns = ['Pap Smear Result', 'Smoking Status', 'STDs History', 'Insurance Covered']
for col in binary_columns:
    if col in clean_df.columns:
        print(f"   Cleaning {col}")
        print(f"      Before: {clean_df[col].unique()}")
        clean_df[col] = clean_df[col].astype(str).str.upper().str.strip()
        print(f"      After: {clean_df[col].unique()}")
        

# 7. Clean screening type
print("\n7. Cleaning Screening Type Last...")
if 'Screening Type Last' in clean_df.columns:
    print(f"   Before: {clean_df['Screening Type Last'].unique()}")
    clean_df['Screening Type Last'] = clean_df['Screening Type Last'].astype(str).str.upper().str.strip()
    print(f"   After: {clean_df['Screening Type Last'].unique()}")
    

# 8. Handle age anomalies
print("\n8. Handling age anomalies...")
print(f"   Age range before: {clean_df['Age'].min()} - {clean_df['Age'].max()}")
clean_df['Age'] = pd.to_numeric(clean_df['Age'], errors='coerce')
# Check for unrealistic ages
unrealistic_ages = clean_df[(clean_df['Age'] < 10) | (clean_df['Age'] > 100)]
if len(unrealistic_ages) > 0:
    print(f"   Found {len(unrealistic_ages)} unrealistic ages")
clean_df['Age'] = clean_df['Age'].clip(lower=10, upper=100)
print(f"   Age range after: {clean_df['Age'].min()} - {clean_df['Age'].max()}")

# 9. Handle First Sexual Activity Age anomalies
print("\n9. Handling First Sexual Activity Age anomalies...")
print(f"   First Sexual Activity Age range before: {clean_df['First Sexual Activity Age'].min()} - {clean_df['First Sexual Activity Age'].max()}")
clean_df['First Sexual Activity Age'] = pd.to_numeric(clean_df['First Sexual Activity Age'], errors='coerce')

# Check for impossible values (first sexual activity age > current age)
# Identify invalid rows
impossible_ages = clean_df['First Sexual Activity Age'] > clean_df['Age']

# Print how many invalid records were found
if impossible_ages.any():
    print(f"   Found {impossible_ages.sum()} records where first sexual activity age > current age")

    # Drop those rows from the DataFrame
    clean_df = clean_df[~impossible_ages]

print("\n10. Handling Sexual Partners...")

# Convert to numeric, non-numeric values become NaN
clean_df['Sexual Partners'] = pd.to_numeric(clean_df['Sexual Partners'], errors='coerce')

# Show range before dropping
print(f"   Sexual Partners range before cleanup: {clean_df['Sexual Partners'].min()} - {clean_df['Sexual Partners'].max()}")

# Drop rows where 'Sexual Partners' is NaN
initial_count = len(clean_df)
clean_df = clean_df.dropna(subset=['Sexual Partners'])
dropped_count = initial_count - len(clean_df)

print(f"   Dropped {dropped_count} rows with missing or invalid 'Sexual Partners' values.")

# 11. Final cleanup - strip all string columns
print("\n11. Final cleanup...")
for col in clean_df.select_dtypes(include='object').columns:
    clean_df[col] = clean_df[col].astype(str).str.strip()

print("\n" + "="*50)
print("CLEANING COMPLETE - SUMMARY:")
print("="*50)
print(f"Original shape: {df.shape}")
print(f"Cleaned shape: {clean_df.shape}")
print(f"Columns after cleaning: {clean_df.columns.tolist()}")

# Show cleaned data quality
print("\nCleaned data quality:")
print("1. HPV Test Result values:", clean_df['HPV Test Result'].unique())
print("2. Regions:", sorted(clean_df['Region'].unique()))
print("3. Missing values after cleaning:")
print(clean_df.isnull().sum())




Loading raw data...
Original dataset shape: (100, 13)

BEFORE CLEANING - Data Quality Issues:
1. Column names:
['Patient ID', 'Age', 'Sexual Partners', 'First Sexual Activity Age', 'HPV Test Result', 'Pap Smear Result', 'Smoking Status', 'STDs History', 'Region', 'Insrance Covered', 'Screening Type Last', 'Recommended Action', 'Unnamed: 12']

2. Sample of problematic data:
HPV Test Result unique values: ['NEGATIVE' 'POSITIVE' 'NEGAGTIVE' 'POSITIVE\n']
Region unique values: ['Pumwani ' 'Kakamega ' 'Machakos' 'Embu ' 'Mombasa' 'NAKURU' 'Loitoktok'
 'Moi ' 'Garissa ' 'Kitale' 'Kakamega' 'Mombasa ' 'Garissa' 'Kericho'
 'Pumwani' 'Kericho ' 'Machakos ' 'Moi' 'Kitale ']
Recommended Action samples:
  - 'REPEAT PAP SMEAR IN 3 YEARS AND FOR HPV VACCINE'
  - 'FOR HPV VACCINE AND SEXUAL EDUCATION'
  - 'FOR HPV VACCINE, LIFESTYLE AND SEXUAL EDUCATION'
  - 'FOR COLPOSCOPY CYTOLOGY AND BIOPSY'
  - 'REPEAT PAP SMEAR IN 3 YEARS'

3. Missing values:
Patient ID                    0
Age                  

Patient ID  Age  Sexual Partners  First Sexual Activity Age  \
0       P0001   18                4                         15   
1       P0002   15                1                         14   
2       P0003   34                1                          9   
3       P0004   52                5                         16   
4       P0005   46                3                         21   
..        ...  ...              ...                        ...   
95      P0096   31                4                         16   
96      P0097   35                5                         11   
97      P0098   35                1                         18   
98      P0099   31                1                         20   
99      P0100   59                6                         22   

   HPV Test Result Pap Smear Result Smoking Status STDs History    Region  \
0         NEGATIVE                N              N            Y   Pumwani   
1         POSITIVE                N              Y            Y  Kakamega   
2         POSITIVE                N              N            Y  Machakos   
3         POSITIVE                N              Y            N      Embu   
4         POSITIVE                N              N            N   Mombasa   
..             ...              ...            ...          ...       ...   
95        POSITIVE                Y              Y            N  Machakos   
96        NEGATIVE                N              Y            N   Mombasa   
97        POSITIVE                N              Y            N   Kericho   
98        POSITIVE                Y              Y            N      Embu   
99        POSITIVE                Y              N            Y    Kitale   

   Insurance Covered Screening Type Last  \
0                  Y           PAP SMEAR   
1                  N             HPV DNA   
2                  N             HPV DNA   
3                  Y             HPV DNA   
4                  N             HPV DNA   
..               ...                 ...   
95                 N           PAP SMEAR   
96                 N           PAP SMEAR   
97                 N             HPV DNA   
98                 Y             HPV DNA   
99                 Y             HPV DNA   

                                 Recommended Action  
0   REPEAT PAP SMEAR IN 3 YEARS AND FOR HPV VACCINE  
1              FOR HPV VACCINE AND SEXUAL EDUCATION  
2              FOR HPV VACCINE AND SEXUAL EDUCATION  
3   FOR HPV VACCINE, LIFESTYLE AND SEXUAL EDUCATION  
4              FOR HPV VACCINE AND SEXUAL EDUCATION  
..                                              ...  
95                  FOR COLPOSCOPY BIOPSY, CYTOLOGY  
96                      REPEAT PAP SMEAR IN 3 YEARS  
97     FOR ANUAL FOLLOW UP AND PAP SMEAR IN 3 YEARS  
98                  FOR COLPOSCOPY BIOPSY, CYTOLOGY  
99           FOR COLPOSCOPY BIOPSY, CYTOLOGY +/-TAH  

[98 rows x 12 columns]

### 3.3: Final Validation and Saving of Cleaned Data

After applying the data cleaning transformations, we perform a final check and save the cleaned dataset for further analysis.

---

#### Steps Included:

1. **Save Cleaned Dataset**  
   - Export the cleaned `DataFrame` to a new CSV file named `'cervical_cancer_cleaned.csv'`.
   - This ensures the cleaned version is preserved and can be reused for further analysis or modeling.

2. **Preview Cleaned Data**  
   - Display the first 5 rows of the cleaned dataset using `.head()` to get a sense of the cleaned structure and content.

3. **Verify Key Columns**  
   - Use a custom `show_unique_values()` function to display unique values for critical categorical features such as:
     - `HPV Test Result`
     - `Region`
     - `Pap Smear Result`
     - `Smoking Status`
     - `STDs History`
     - `Insurance Covered`
     - `Screening Type Last`
   - This helps confirm that cleaning steps were effective in removing typos, inconsistencies, and formatting issues.

4. **Column-Wise Value Counts**  
   - The `print_all_value_counts()` function prints value counts for all columns in the dataset (including `NaNs`).
   - This helps detect any remaining anomalies and understand the distribution of each variable, which is critical for:
     - Identifying imbalances
     - Planning for encoding (e.g., one-hot, label)
     - Feature engineering

---

#### Conclusion:

-  The cleaned dataset has been saved successfully.
-  Categorical columns show standardized and consistent values.
-  There is no significant presence of missing or malformed data.
-  The dataset is now ready for the **Exploratory Data Analysis (EDA)** phase.


In [ ]:
# Save cleaned data
clean_df.to_csv('cervical_cancer_cleaned.csv', index=False)
print("\nCleaned data saved as 'cervical_cancer_cleaned.csv'")

# Display first few rows
print("\nFirst 5 rows of cleaned data:")
print(clean_df.head())

# Show unique values for key columns to verify cleaning
print("\n" + "="*50)
print("VERIFICATION OF CLEANED DATA:")
print("="*50)

def show_unique_values(df, columns):
    for col in columns:
        if col in df.columns:
            print(f"\n{col}:")
            unique_vals = df[col].unique()
            if len(unique_vals) <= 10:
                print(f"  Values: {unique_vals}")
            else:
                print(f"  Total unique values: {len(unique_vals)}")
                print(f"  Sample: {unique_vals[:10]}")

verification_columns = ['HPV Test Result', 'Region', 'Pap Smear Result', 'Smoking Status', 
                       'STDs History', 'Insurance Covered', 'Screening Type Last']
show_unique_values(clean_df, verification_columns)


#Preview results
def print_all_value_counts(df):
    for column in df.columns:
        print(f"--- Value Counts for: {column} ---")
        print(df[column].value_counts(dropna=False))  # include NaNs in the count
        print("\n")

# Print unique values to check validity & quality:
print_all_value_counts(clean_df)

print("\n✅ Data cleaning completed successfully!")
print("✅ Ready for next step: Exploratory Data Analysis")
clean_df

  Patient ID  Age  Sexual Partners  First Sexual Activity Age HPV Test Result  \
0      P0001   18                4                         15        NEGATIVE   
1      P0002   15                1                         14        POSITIVE   
2      P0003   34                1                          9        POSITIVE   
3      P0004   52                5                         16        POSITIVE   
4      P0005   46                3                         21        POSITIVE   

  Pap Smear Result Smoking Status STDs History     Region Insrance Covered  \
0                N              N            Y   Pumwani                 Y   
1                N              Y            Y  Kakamega                 N   
2                N              N            Y   Machakos                N   
3                N              Y            N      Embu                 Y   
4                N              N            N    Mombasa                N   

  Screening Type Last                       

### 3.4: Advanced Cleaning of "Recommended Action" Column using Fuzzy Matching and Keyword Rules

This step addresses one of the most inconsistent and typo-ridden columns in the dataset — `Recommended Action`. It applies a structured, rule-based fuzzy matching approach to standardize all entries against a predefined list of **canonical (correct) values**.

---

#### Step Breakdown:

1. **Define Canonical Values**
   - These are the 100% correct, clean strings that we expect the column to match.
   - They cover all valid and normalized options like:
     - `"FOR COLPOSCOPY, BIOPSY, AND CYTOLOGY"`
     - `"REPEAT PAP SMEAR IN 3 YEARS"`
     - `"FOR HPV VACCINE, LIFESTYLE, AND SEXUAL EDUCATION"`

2. **Set up Typo Mapping**
   - Common misspellings like:
     - `"BIOSPY"` → `"BIOPSY"`
     - `"COLOSCOPY"` → `"COLPOSCOPY"`
     - `"ANUAL"` → `"ANNUAL"`
   - These are corrected **before matching** to improve accuracy.

3. **Strict Fuzzy Matching Function**
   - A custom function `clean_column_strict()` is used to:
     - Standardize text to uppercase.
     - Correct known typos.
     - Extract and match **critical keywords** (like `HPV`, `CYTOLOGY`, `BIOPSY`, etc.).
     - Use **fuzzy string matching** (via `fuzzywuzzy`) to find the closest canonical value.
     - Ensure matches are **only accepted** if the keywords from the raw and canonical string are identical. This prevents incorrect but similar matches.

4. **Apply Cleaning**
   - The function is called with the column name, canonical list, and threshold (80% match confidence).
   - It returns:
     - A cleaned version of the column
     - A confidence score
     - A status flag (`OK` or `NEEDS MANUAL REVIEW`)

5. **Manual Review Report**
   - Any rows where the fuzzy match is below the confidence threshold (or keyword sets don’t match) are flagged for **manual review**.

6. **Save the Final Data**
   - The updated `DataFrame` with cleaned `Recommended Action` is exported to an Excel file:  
      `final_cleaned.xlsx`

---

#### Why This Approach?

- Fuzzy matching alone can lead to incorrect substitutions when two strings are similar but mean different things.
- By combining **keyword presence checks** and **typo correction**, we ensure data is matched **both intelligently and safely**.
- This is particularly useful in messy real-world healthcare data.

---

#### Outcome

- The `Recommended Action` column is now fully standardized.
- Any unmatched or low-confidence rows are clearly flagged for manual inspection.
- The dataset is fully cleaned and saved, ready for Exploratory Data Analysis (EDA).


In [ ]:
# This list contains the final, correct values to use.
canonical_values = [
    "REPEAT PAP SMEAR IN 3 YEARS",
    "FOR COLPOSCOPY, BIOPSY, AND CYTOLOGY",
    "FOR PAP SMEAR",
    "FOR HPV VACCINE AND SEXUAL EDUCATION",
    "FOR ANNUAL FOLLOW-UP AND PAP SMEAR IN 3 YEARS",
    "REPEAT PAP SMEAR IN 3 YEARS AND FOR HPV VACCINE",
    "FOR COLPOSCOPY, BIOPSY, AND CYTOLOGY (TAH NOT RECOMMENDED)",
    "FOR BIOPSY AND CYTOLOGY (TAH NOT RECOMMENDED)",
    "FOR HPV VACCINE, LIFESTYLE, AND SEXUAL EDUCATION",
    "FOR COLPOSCOPY, BIOPSY, CYTOLOGY +/- TAH",
    "FOR COLPOSCOPY, CYTOLOGY, THEN LASER THERAPY",
    "FOR REPEAT HPV TESTING ANNUALLY AND PAP SMEAR IN 3 YEARS",
    "FOR LASER THERAPY"
]

# --- 2. Configuration for the Cleaner ---
TYPO_MAP = {
    'COLPOSOCPY': 'COLPOSCOPY', 'COLPOSCPY': 'COLPOSCOPY', 'COLOSCOPY': 'COLPOSCOPY',
    'BIOSPY': 'BIOPSY', 'BIOSY': 'BIOPSY', 'ANUAL': 'ANNUAL', 'VACCINATION': 'VACCINE'
}
CRITICAL_KEYWORDS = {
    'COLPOSCOPY', 'BIOPSY', 'CYTOLOGY', 'TAH', 'HPV', 'LASER', 'LIFESTYLE', 'ANNUALLY'
}

# --- 3. The Definitive Data Cleaning Function ---
def clean_column_strict(df, column_to_clean, canonical_list, threshold=80):
    """
    Cleans a DataFrame column using fuzzy matching combined with a STRICT keyword set equality rule.
    """
    if column_to_clean not in df.columns:
        raise ValueError(f"Column '{column_to_clean}' not found in the DataFrame.")

    def preprocess_and_get_keywords(text, typo_map):
        if not isinstance(text, str): return "", set()
        # Standardize to uppercase and correct typos
        processed_text = text.upper()
        for wrong, right in typo_map.items():
            processed_text = re.sub(r'\b' + wrong + r'\b', right, processed_text)
        
        # Extract the set of critical keywords
        keywords = {word for word in CRITICAL_KEYWORDS if word in processed_text}
        return processed_text, keywords

    cleaned_data = []

    # Pre-calculate keywords for the canonical list for efficiency
    canonical_keywords = {val: preprocess_and_get_keywords(val, {})[1] for val in canonical_list}

    for index, row in df.iterrows():
        raw_string = row[column_to_clean]
        
        if not isinstance(raw_string, str) or not raw_string.strip():
            cleaned_data.append((raw_string, 0, 'NO_DATA_PROVIDED'))
            continue

        processed_string, original_keywords = preprocess_and_get_keywords(raw_string, TYPO_MAP)
        
        # Get the top few potential matches instead of just one
        top_matches = process.extract(processed_string, canonical_list, limit=5)

        best_valid_match = None
        
        for potential_match, score in top_matches:
            if score < threshold:
                break # No need to check further if scores are too low

            # The CRITICAL check: The keyword sets must be identical
            if original_keywords == canonical_keywords[potential_match]:
                best_valid_match = (potential_match, score)
                break # Found the best possible valid match, stop searching

        if best_valid_match:
            final_value, final_score = best_valid_match
            status = 'OK'
        else:
            final_value = raw_string
            final_score = top_matches[0][1] if top_matches else 0 # Show score of best (but invalid) match
            status = 'NEEDS MANUAL REVIEW'

        cleaned_data.append((final_value, final_score, status))

    result_df = pd.DataFrame(cleaned_data, index=df.index, columns=[f'{column_to_clean}_Cleaned', f'{column_to_clean}_Confidence', f'{column_to_clean}_Status'])
    return df.join(result_df)

# main_df = pd.DataFrame(data, columns=['Recommended Action'])


# 2. Call the function to clean the specified column
# You provide your DataFrame, the column name, the list of correct values, and the confidence threshold.
final_df = clean_column_strict(
    df=clean_df,
    column_to_clean='Recommended Action',
    canonical_list=canonical_values,
    threshold=80
)

# 3. Display the results
print("--- Full Data Cleaning Results ---")
# Use .to_string() to ensure all columns are displayed without truncation
#print(final_df.to_string())

print("\n" + "="*80 + "\n")

# 4. Display only the rows that need human attention
print("--- Actions Flagged for Manual Review (Confidence < 80%) ---")
manual_review_df = final_df[final_df['Recommended Action_Status'] == 'NEEDS MANUAL REVIEW']

if manual_review_df.empty:
    print("No actions require manual review. All items met the 80% confidence threshold.")
else:
    print(manual_review_df[['Recommended Action', 'Recommended Action_Cleaned', 'Recommended Action_Confidence']].to_string())


print(f"   After: {final_df['Recommended Action_Cleaned'].unique()}")
final_df
final_df.to_excel("final_cleaned.xlsx",index=False)


  Patient ID  Age  Sexual Partners  First Sexual Activity Age HPV Test Result  \
0      P0001   18                4                         15        NEGATIVE   
1      P0002   15                1                         14        POSITIVE   
2      P0003   34                1                          9        POSITIVE   
3      P0004   52                5                         16        POSITIVE   
4      P0005   46                3                         21        POSITIVE   

  Pap Smear Result Smoking Status STDs History     Region Insrance Covered  \
0                N              N            Y   Pumwani                 Y   
1                N              Y            Y  Kakamega                 N   
2                N              N            Y   Machakos                N   
3                N              Y            N      Embu                 Y   
4                N              N            N    Mombasa                N   

  Screening Type Last                       

###  3.5: Final Cleanup – Dropping Redundant Columns

In this step, we perform a final cleanup to remove columns that are no longer needed for analysis or modeling. These include metadata and intermediate columns from previous cleaning stages.

---

####  Breakdown:

1. **Load Final Cleaned Data**
   - Reads the Excel file `final_cleaned.xlsx`, which contains the dataset with standardized `Recommended Action` and other cleaned columns.

2. **Remove Redundant Columns**
   - The following columns are dropped:
     - `'Recommended Action'`: Original raw column (replaced by the cleaned version).
     - `'Recommended Action_Confidence'`: Confidence score from the fuzzy matching process (used for diagnostics).
     - `'Recommended Action_Status'`: Flag for manual review (used during cleaning).
     - `'Patient ID'`: Identifier column not relevant to statistical or ML modeling.

3. **Save the Resulting Dataset**
   - Exports the trimmed dataset as `data_final.xlsx`, which now contains only relevant and clean features for **Exploratory Data Analysis (EDA)** or modeling.

---

 This marks the completion of the data cleaning phase.  
 The next step is **Exploratory Data Analysis (EDA)** to understand the structure, relationships, and patterns in the data.


In [88]:
#Load the file
file_path = 'final_cleaned.xlsx'
df = pd.read_excel(file_path)
print("1. Removing unnecessary columns...")
columns_to_drop = ['Recommended Action','Recommended Action_Confidence','Recommended Action_Status','Patient ID']
for col in columns_to_drop:
    if col in df.columns:
        df.drop(columns=[col], inplace=True)
df.to_excel('data_final.xlsx', index=False)

--- Full Data Cleaning Results ---


--- Actions Flagged for Manual Review (Confidence < 80%) ---
No actions require manual review. All items met the 80% confidence threshold.
   After: ['REPEAT PAP SMEAR IN 3 YEARS AND FOR HPV VACCINE'
 'FOR HPV VACCINE AND SEXUAL EDUCATION'
 'FOR HPV VACCINE, LIFESTYLE, AND SEXUAL EDUCATION'
 'FOR COLPOSCOPY, BIOPSY, AND CYTOLOGY' 'REPEAT PAP SMEAR IN 3 YEARS'
 'FOR COLPOSCOPY, BIOPSY, CYTOLOGY +/- TAH'
 'FOR BIOPSY AND CYTOLOGY (TAH NOT RECOMMENDED)'
 'FOR ANNUAL FOLLOW-UP AND PAP SMEAR IN 3 YEARS' 'FOR PAP SMEAR'
 'FOR COLPOSCOPY, BIOPSY, AND CYTOLOGY (TAH NOT RECOMMENDED)'
 'FOR LASER THERAPY' 'FOR COLPOSCOPY, CYTOLOGY, THEN LASER THERAPY'
 'FOR REPEAT HPV TESTING ANNUALLY AND PAP SMEAR IN 3 YEARS']



# STEP 1: STANDARDIZING TARGET COLUMN VALUES(RECOMMENDATION)
## Goal: Balance the data on the target column so that the model is not biased to a certain recommendation.

We'll do:

* Define our cannonical values.

* Using confidence score and fuzzy matching for the standardization.Values that do not meet confoidence score are flagged for manual checking



In [88]:
# This list contains the final, correct values to use.
canonical_values = [
    "REPEAT PAP SMEAR IN 3 YEARS",
    "FOR COLPOSCOPY, BIOPSY, AND CYTOLOGY",
    "FOR PAP SMEAR",
    "FOR HPV VACCINE AND SEXUAL EDUCATION",
    "FOR ANNUAL FOLLOW-UP AND PAP SMEAR IN 3 YEARS",
    "REPEAT PAP SMEAR IN 3 YEARS AND FOR HPV VACCINE",
    "FOR COLPOSCOPY, BIOPSY, AND CYTOLOGY (TAH NOT RECOMMENDED)",
    "FOR BIOPSY AND CYTOLOGY (TAH NOT RECOMMENDED)",
    "FOR HPV VACCINE, LIFESTYLE, AND SEXUAL EDUCATION",
    "FOR COLPOSCOPY, BIOPSY, CYTOLOGY +/- TAH",
    "FOR COLPOSCOPY, CYTOLOGY, THEN LASER THERAPY",
    "FOR REPEAT HPV TESTING ANNUALLY AND PAP SMEAR IN 3 YEARS",
    "FOR LASER THERAPY"
]

# --- 2. Configuration for the Cleaner ---
TYPO_MAP = {
    'COLPOSOCPY': 'COLPOSCOPY', 'COLPOSCPY': 'COLPOSCOPY', 'COLOSCOPY': 'COLPOSCOPY',
    'BIOSPY': 'BIOPSY', 'BIOSY': 'BIOPSY', 'ANUAL': 'ANNUAL', 'VACCINATION': 'VACCINE'
}
CRITICAL_KEYWORDS = {
    'COLPOSCOPY', 'BIOPSY', 'CYTOLOGY', 'TAH', 'HPV', 'LASER', 'LIFESTYLE', 'ANNUALLY'
}

# --- 3. The Definitive Data Cleaning Function ---
def clean_column_strict(df, column_to_clean, canonical_list, threshold=80):
    """
    Cleans a DataFrame column using fuzzy matching combined with a STRICT keyword set equality rule.
    """
    if column_to_clean not in df.columns:
        raise ValueError(f"Column '{column_to_clean}' not found in the DataFrame.")

    def preprocess_and_get_keywords(text, typo_map):
        if not isinstance(text, str): return "", set()
        # Standardize to uppercase and correct typos
        processed_text = text.upper()
        for wrong, right in typo_map.items():
            processed_text = re.sub(r'\b' + wrong + r'\b', right, processed_text)
        
        # Extract the set of critical keywords
        keywords = {word for word in CRITICAL_KEYWORDS if word in processed_text}
        return processed_text, keywords

    cleaned_data = []

    # Pre-calculate keywords for the canonical list for efficiency
    canonical_keywords = {val: preprocess_and_get_keywords(val, {})[1] for val in canonical_list}

    for index, row in df.iterrows():
        raw_string = row[column_to_clean]
        
        if not isinstance(raw_string, str) or not raw_string.strip():
            cleaned_data.append((raw_string, 0, 'NO_DATA_PROVIDED'))
            continue

        processed_string, original_keywords = preprocess_and_get_keywords(raw_string, TYPO_MAP)
        
        # Get the top few potential matches instead of just one
        top_matches = process.extract(processed_string, canonical_list, limit=5)

        best_valid_match = None
        
        for potential_match, score in top_matches:
            if score < threshold:
                break # No need to check further if scores are too low

            # The CRITICAL check: The keyword sets must be identical
            if original_keywords == canonical_keywords[potential_match]:
                best_valid_match = (potential_match, score)
                break # Found the best possible valid match, stop searching

        if best_valid_match:
            final_value, final_score = best_valid_match
            status = 'OK'
        else:
            final_value = raw_string
            final_score = top_matches[0][1] if top_matches else 0 # Show score of best (but invalid) match
            status = 'NEEDS MANUAL REVIEW'

        cleaned_data.append((final_value, final_score, status))

    result_df = pd.DataFrame(cleaned_data, index=df.index, columns=[f'{column_to_clean}_Cleaned', f'{column_to_clean}_Confidence', f'{column_to_clean}_Status'])
    return df.join(result_df)

# main_df = pd.DataFrame(data, columns=['Recommended Action'])


# 2. Call the function to clean the specified column
# You provide your DataFrame, the column name, the list of correct values, and the confidence threshold.
final_df = clean_column_strict(
    df=clean_df,
    column_to_clean='Recommended Action',
    canonical_list=canonical_values,
    threshold=80
)

# 3. Display the results
print("--- Full Data Cleaning Results ---")
# Use .to_string() to ensure all columns are displayed without truncation
#print(final_df.to_string())

print("\n" + "="*80 + "\n")

# 4. Display only the rows that need human attention
print("--- Actions Flagged for Manual Review (Confidence < 80%) ---")
manual_review_df = final_df[final_df['Recommended Action_Status'] == 'NEEDS MANUAL REVIEW']

if manual_review_df.empty:
    print("No actions require manual review. All items met the 80% confidence threshold.")
else:
    print(manual_review_df[['Recommended Action', 'Recommended Action_Cleaned', 'Recommended Action_Confidence']].to_string())


print(f"   After: {final_df['Recommended Action_Cleaned'].unique()}")
final_df
final_df.to_excel("final_cleaned.xlsx",index=False)



--- Full Data Cleaning Results ---


--- Actions Flagged for Manual Review (Confidence < 80%) ---
No actions require manual review. All items met the 80% confidence threshold.
   After: ['REPEAT PAP SMEAR IN 3 YEARS AND FOR HPV VACCINE'
 'FOR HPV VACCINE AND SEXUAL EDUCATION'
 'FOR HPV VACCINE, LIFESTYLE, AND SEXUAL EDUCATION'
 'FOR COLPOSCOPY, BIOPSY, AND CYTOLOGY' 'REPEAT PAP SMEAR IN 3 YEARS'
 'FOR COLPOSCOPY, BIOPSY, CYTOLOGY +/- TAH'
 'FOR BIOPSY AND CYTOLOGY (TAH NOT RECOMMENDED)'
 'FOR ANNUAL FOLLOW-UP AND PAP SMEAR IN 3 YEARS' 'FOR PAP SMEAR'
 'FOR COLPOSCOPY, BIOPSY, AND CYTOLOGY (TAH NOT RECOMMENDED)'
 'FOR LASER THERAPY' 'FOR COLPOSCOPY, CYTOLOGY, THEN LASER THERAPY'
 'FOR REPEAT HPV TESTING ANNUALLY AND PAP SMEAR IN 3 YEARS']


In [89]:
#Load the file
file_path = 'final_cleaned.xlsx'
df = pd.read_excel(file_path)
print("1. Removing unnecessary columns...")
columns_to_drop = ['Recommended Action','Recommended Action_Confidence','Recommended Action_Status','Patient ID']
for col in columns_to_drop:
    if col in df.columns:
        df.drop(columns=[col], inplace=True)
df.to_excel('data_final.xlsx', index=False)

1. Removing unnecessary columns...



# STEP 2: DATA SPLITTING
## Goal:  Split the dataset to get the portion that goes into training and the portion that goes into testing

We'll do:

* Pre-split Fix.We notice that in the dataset,some critical columns have only one instance of it.Here we will use simple oversampling by manually duplicating these classes with only one instance.
* stratified splitting on the data to ensure equal amount of percentage goes into both training and testing.
* hot-encode remaining columns
  # STEP 2B
  ## Goal :Define class weights based on domain knowledge(Medical) to be fed into the data for use during training .





In [95]:
# --- 1. Load Your Data ---
# Let's create a sample DataFrame that mimics your situation

data = pd.read_excel('data_final.xlsx')
df = pd.DataFrame(data)



# --- 2. DEFINE TARGET and SEPARATE X and y ---
target_column = 'Recommended Action_Cleaned'
y = df[target_column]
X = df.drop(columns=[target_column])


# --- 3. PRE-SPLIT STEP: LABEL ENCODE THE TARGET COLUMN ---
# This is required so that the 'stratify' parameter can work with numeric labels.
print("\n--- 3. Label Encoding Target Column (y) ---")
le = LabelEncoder()
y_encoded = le.fit_transform(y)
print("Target column converted to numeric labels.")
print("-" * 60)


# --- 4. PRE-SPLIT STEP: HANDLE SINGLE-INSTANCE CLASSES for STRATIFICATION ---
# Stratify needs at least 2 members per class. We duplicate the singletons.
print("\n--- 4. Checking for and Fixing Single-Instance Classes ---")
class_counts = pd.Series(y_encoded).value_counts()
single_instance_classes = class_counts[class_counts < 2].index

X_fixed = X.copy()
y_fixed = y_encoded.copy()

if len(single_instance_classes) > 0:
    print(f"Found {len(single_instance_classes)} classes with only one sample. Duplicating them...")
    
    indices_to_duplicate = pd.Series(y_encoded).isin(single_instance_classes)
    
    X_to_duplicate = X[indices_to_duplicate]
    y_to_duplicate = y_encoded[indices_to_duplicate]
    
    X_fixed = pd.concat([X, X_to_duplicate], ignore_index=True)
    y_fixed = pd.concat([pd.Series(y_encoded), pd.Series(y_to_duplicate)], ignore_index=True).values
else:
    print("No single-instance classes found that would break stratification.")

print(f"Shape before fix: {X.shape}. Shape after fix: {X_fixed.shape}.")
print("-" * 60)

    
# --- 5. SPLIT THE DATA (CRITICAL LEAKAGE-PREVENTION STEP) ---
# We split the 'fixed' data. X_fixed STILL CONTAINS THE TEXT/CATEGORICAL COLUMNS.
print("\n--- 5. Splitting Data into Training and Testing Sets ---")
X_train, X_test, y_train, y_test = train_test_split(
    X_fixed, y_fixed, 
    test_size=0.2,       # 20% for testing
    random_state=42,     # For reproducibility
    stratify=y_fixed     # Ensure class distribution is similar
)
print("Data successfully split.")
print(f"Training set size: {len(X_train)} rows")
print(f"Test set size: {len(X_test)} rows")
print("-" * 60)


# --- 6. ONE-HOT ENCODE THE FEATURE COLUMNS (POST-SPLIT) ---
# We learn the encoding schema ONLY from the training data to prevent data leakage.
print("\n--- 6. One-Hot Encoding Categorical Features ---")
# Automatically identify which columns are numeric and which are categorical from the training set
numeric_features = X_train.select_dtypes(include='number').columns
categorical_features = X_train.select_dtypes(include='object').columns

print(f"Identified Numeric Features: {list(numeric_features)}")
print(f"Identified Categorical Features: {list(categorical_features)}")

# Create the encoder
encoder = OneHotEncoder(handle_unknown='ignore', sparse_output=False)

# FIT the encoder ONLY on the TRAINING data's categorical columns
encoder.fit(X_train[categorical_features])

# TRANSFORM both the training and testing data's categorical columns
X_train_encoded = pd.DataFrame(
    encoder.transform(X_train[categorical_features]),
    columns=encoder.get_feature_names_out(categorical_features),
    index=X_train.index
)
X_test_encoded = pd.DataFrame(
    encoder.transform(X_test[categorical_features]),
    columns=encoder.get_feature_names_out(categorical_features),
    index=X_test.index
)

# Reset index on all parts to ensure clean concatenation
X_train_numeric_reset = X_train[numeric_features].reset_index(drop=True)
X_test_numeric_reset = X_test[numeric_features].reset_index(drop=True)
X_train_encoded_reset = X_train_encoded.reset_index(drop=True)
X_test_encoded_reset = X_test_encoded.reset_index(drop=True)

# Combine encoded categorical columns with the original numeric columns
X_train_processed = pd.concat([X_train_numeric_reset, X_train_encoded_reset], axis=1)
X_test_processed = pd.concat([X_test_numeric_reset, X_test_encoded_reset], axis=1)

print("\nFeatures have been successfully encoded.")
print(f"Shape of processed training features: {X_train_processed.shape}")
print(f"Shape of processed testing features: {X_test_processed.shape}")
print("-" * 60)

print("\n--- 5. Manually Defining Class Weights Based on Clinical Importance ---")

# First, let's create a mapping from class names to their encoded labels for clarity
label_mapping = dict(zip(le.classes_, le.transform(le.classes_)))
print("Class Label Mapping:")
print(label_mapping)

# ===================================================================================
# THIS IS WHERE YOU APPLY YOUR DOMAIN KNOWLEDGE.
# You create the dictionary from scratch. Higher weight = higher penalty for getting it wrong.
# You MUST provide a weight for every class label present in y_train.
# ===================================================================================
manual_weights_dict = {
    label_mapping['REPEAT PAP SMEAR IN 3 YEARS AND FOR HPV VACCINE']: 1.0, # Majority/common class
    label_mapping['FOR HPV VACCINE AND SEXUAL EDUCATION']: 1.2, # Slightly less common
    label_mapping['FOR COLPOSCOPY, BIOPSY, AND CYTOLOGY']: 20.0, # IMPORTANT - High penalty
    label_mapping['FOR LASER THERAPY']: 30.0, # CRITICALLY IMPORTANT - Very high penalty
    label_mapping['FOR COLPOSCOPY, BIOPSY, CYTOLOGY +/- TAH']: 25.0, # IMPORTANT - High penalty
    label_mapping["REPEAT PAP SMEAR IN 3 YEARS"]:10.0,
    label_mapping["FOR PAP SMEAR"]:5.0,
    label_mapping["FOR ANNUAL FOLLOW-UP AND PAP SMEAR IN 3 YEARS"]:20.0,
    label_mapping["FOR HPV VACCINE, LIFESTYLE, AND SEXUAL EDUCATION"]:1.5,
    label_mapping["FOR COLPOSCOPY, CYTOLOGY, THEN LASER THERAPY"] :25.0,
    label_mapping["FOR REPEAT HPV TESTING ANNUALLY AND PAP SMEAR IN 3 YEARS"]:25.0,
    label_mapping['FOR COLPOSCOPY, BIOPSY, AND CYTOLOGY (TAH NOT RECOMMENDED)']: 20.0,
    label_mapping['FOR BIOPSY AND CYTOLOGY (TAH NOT RECOMMENDED)']: 20.0
}

print("\nCustom Weights Defined:")
print(manual_weights_dict)
print("-" * 60)






--- 3. Label Encoding Target Column (y) ---
Target column converted to numeric labels.
------------------------------------------------------------

--- 4. Checking for and Fixing Single-Instance Classes ---
Found 6 classes with only one sample. Duplicating them...
Shape before fix: (98, 10). Shape after fix: (104, 10).
------------------------------------------------------------

--- 5. Splitting Data into Training and Testing Sets ---
Data successfully split.
Training set size: 83 rows
Test set size: 21 rows
------------------------------------------------------------

--- 6. One-Hot Encoding Categorical Features ---
Identified Numeric Features: ['Age', 'Sexual Partners', 'First Sexual Activity Age']
Identified Categorical Features: ['HPV Test Result', 'Pap Smear Result', 'Smoking Status', 'STDs History', 'Region', 'Insurance Covered', 'Screening Type Last']

Features have been successfully encoded.
Shape of processed training features: (83, 27)
Shape of processed testing features: 


# STEP 3: Random Forest Classifier
## Goal: Understand the features which have more influence on the target column. The columns with the most correlation will then be used to train the model.


In [ ]:



# --- Stage A: Calculate Feature Importance and Select Features (on training data) ---
print("\n--- A. Calculating Feature Importance on Training Data ONLY ---")

# We train a preliminary RandomForest model to get feature importances.
# It's good practice to use the class weights here too so the importance calculation
# is not biased towards features that are only useful for the majority class.
feature_selector_model = RandomForestClassifier(
    random_state=42,
    class_weight=manual_weights_dict
)
feature_selector_model.fit(X_train_processed, y_train)

# Get the importance scores
importances = feature_selector_model.feature_importances_
feature_names = X_train_processed.columns
importance_df = pd.DataFrame({'feature': feature_names, 'importance': importances}).sort_values(by='importance', ascending=False)

print("Top 8 most important features (learned from training data):")
print(importance_df.head(8))

# --- Stage B: Perform the Feature Selection ---
print("\n--- B. Selecting Features Based on Importance ---")

# We can use a threshold (e.g., select features with importance > 0.01)
# Or we can use a helper like SelectFromModel to automatically select them.
# 'threshold="median"' will select all features with importance greater than the median.
selector = SelectFromModel(feature_selector_model, threshold='median', prefit=True)

# Use the selector to get the final versions of our datasets
X_train_final = selector.transform(X_train_processed)
X_test_final = selector.transform(X_test_processed) # Apply same transformation to test set

print(f"\nOriginal number of features: {X_train_processed.shape[1]}")
print(f"Number of features selected: {X_train_final.shape[1]}")
print("-" * 60)


In [ ]:
# 1
import seaborn as sns
import matplotlib.pyplot as plt

# Check class balance for HPV
sns.countplot(data=clean_df, x='HPV Test Result')
plt.title('HPV Test Result Distribution')
plt.show()

# Check class balance for Pap Smear
sns.countplot(data=clean_df, x='Pap Smear Result')
plt.title('Pap Smear Result Distribution')
plt.show()

# Age distribution
plt.figure(figsize=(10, 4))
sns.histplot(data=clean_df, x='Age', bins=20, kde=True)
plt.title('Age Distribution')
plt.show()

# Correlation matrix for numeric columns
numeric_cols = clean_df.select_dtypes(include=['int64', 'float64']).columns
plt.figure(figsize=(12, 8))
sns.heatmap(clean_df[numeric_cols].corr(), annot=True, cmap='coolwarm')
plt.title('Correlation Matrix')
plt.show()


In Exploratory Data Analysis (EDA), We focus on key target variables.

In cervical cancer prediction, we treat this as a binary classification:

Does this patient show signs that are likely to lead to cervical cancer?

We use:

Recommended Action → Could be used as proxy for diagnosis
HPV Test Result
Pap Smear Result
Screening Type Last

We’ll use these to understand who’s at risk.

In [ ]:
# 2.Show distributions of key category values defined above

import matplotlib.pyplot as plt
import seaborn as sns

# Set visual style
sns.set(style='whitegrid')

# HPV Test Result distribution
plt.figure(figsize=(6,4))
sns.countplot(data=clean_df, x='HPV Test Result', palette='Set2')
plt.title('Distribution of HPV Test Results')
plt.ylabel('Count')
plt.show()

# Pap Smear Result distribution
plt.figure(figsize=(6,4))
sns.countplot(data=clean_df, x='Pap Smear Result', palette='Set3')
plt.title('Distribution of Pap Smear Results')
plt.ylabel('Count')
plt.show()

# Recommended Action distribution (top 10)
plt.figure(figsize=(10,6))
top_actions = clean_df['Recommended Action'].value_counts().nlargest(10)
sns.barplot(y=top_actions.index, x=top_actions.values, palette='coolwarm')
plt.title('Top 10 Recommended Actions')
plt.xlabel('Count')
plt.ylabel('Recommended Action')
plt.show()



In [ ]:
# List of important numerical features
num_features = ['Age', 'Sexual Partners', 'First Sexual Activity Age']

# Plot histograms
clean_df[num_features].hist(bins=20, figsize=(12, 6), color='skyblue')
plt.suptitle('Distribution of Key Numeric Features')
plt.show()


In [ ]:
# Correlation Heatmap to check which features are correlated with others.

# only numerical columns
numeric_cols = clean_df.select_dtypes(include=['int64', 'float64']).columns

plt.figure(figsize=(12, 8))
sns.heatmap(clean_df[numeric_cols].corr(), annot=True, cmap='coolwarm', linewidths=0.5)
plt.title('Correlation Between Numerical Features')
plt.show()


# STEP 3: Modeling and Pre-Training Phase

##  Goal:
Build and evaluate a predictive model using your **cleaned**, **preprocessed** dataset.  
This phase helps us learn patterns from the data and **predict High Risk patients** based on clinical features.

---

## Summary Table of Key Steps

| Step                        | What it Means                                                                 | What You Do                                                   |
|-----------------------------|-------------------------------------------------------------------------------|---------------------------------------------------------------|
| 1. Feature & Target Selection | Choose what we want to predict (target) and the data used to predict it (features). | `X = datav_clean[features]` <br> `y = datav_clean['High Risk']` |
| 2. Train-Test Splitting    | Split the dataset into training and test sets for fair model evaluation.    | `train_test_split(X, y, test_size=0.2, random_state=0)`       |
| 3. Scaling / Encoding (if needed) | Standardize numeric values to a common scale (important for some models like logistic regression). | `StandardScaler()` or `MinMaxScaler()` if model needs it      |
| 4. Model Selection         | Choose a model suitable for classification.                                 | `DecisionTreeClassifier()`, `LogisticRegression()`, etc.      |
| 5. Model Training         | Feed the training data into the model so it can learn patterns.            | `model.fit(X_train, y_train)`                                 |
| 6. Evaluation              | Check the model’s accuracy, precision, recall, etc. using test data.        | `confusion_matrix`, `classification_report`, `accuracy_score` |
| 7. Interpretation          | Understand feature importance and model behavior.                          | `model.feature_importances_` or tree plots                    |

---

## Output of This Phase:
- A trained and evaluated model that can predict if a patient is **High Risk** for cervical cancer.
- Metrics to judge how well the model performs.
- A ranked list of which features are most important in making predictions (e.g., age, smoking, screening type).

---


In [ ]:

# Clean messy strings (e.g., " yes " → "YES")

#Maps them to numbers
#Prepares the dataframe to work seamlessly with your existing risk-flagging logic.

# Ensure all relevant columns in clean_df are encoded before making a copy
clean_df

clean_df['Pap Smear Result'] = clean_df['Pap Smear Result'].astype(str).str.upper().str.strip().replace({
    'N': 0, 'Y': 1
})

clean_df['Smoking Status'] = clean_df['Smoking Status'].astype(str).str.upper().str.strip().replace({
    'N': 0, 'Y': 1, 'TRUE': 1, 'FALSE': 0
})

clean_df['STDs History'] = clean_df['STDs History'].astype(str).str.upper().str.strip().replace({
    'N': 0, 'Y': 1, 'TRUE': 1, 'FALSE': 0
})

clean_df['Insurance Covered'] = clean_df['Insurance Covered'].astype(str).str.upper().str.strip().replace({
    'N': 0, 'Y': 1
})

clean_df


In [ ]:
#Because the cervical cancer dataset doesn’t contain a direct diagnosis column like "has cancer = yes/no", 
#we are creating a new feature called High Risk. 
# we use real clinical clues (like positive test results or medical recommendations) to determine whether a patient is high risk.
# This new High Risk column becomes your true target — better than using only HPV Test Result

# 1. Make a fresh copy
datav = clean_df.copy()

# 2. Encode relevant binary columns. 'YES' = 1 (indicating presence)'NO' = 0 (indicating absence)
datav['HPV Test Result'] = datav['HPV Test Result'].map({'NEGATIVE': 0, 'POSITIVE': 1})

# 3. Handle Screening Type Last → create new binary feature: Screening Type Suspicious, first clean this column:
# Replace missing values (NaN) with 'UNKNOWN'
# Standardize capitalization (so 'hpv', 'HPV ', 'Hpv' all become 'HPV')

#Remove leading/trailing spaces
datav['Screening Type Last'] = datav['Screening Type Last'].fillna('UNKNOWN').str.upper().str.strip()
datav['Screening Type Suspicious'] = datav['Screening Type Last'].apply(
    lambda x: 1 if any(word in x for word in ['COLPOSCOPY', 'PAP', 'HPV']) else 0
)

# 4. Create HIGH RISK target variable ie
# We are flagging a patient as high risk if any of these are true:

# Their HPV test is positive
# Their Pap smear is abnormal
# Their last screening type was serious (like HPV or Colposcopy)
# Their doctor recommended a follow-up action like biopsy or colposcopy

datav['High Risk'] = 0  # default

# High risk if HPV positive AND Pap Smear positive
datav.loc[(datav['HPV Test Result'] == 1) & (datav['Pap Smear Result'] == 1), 'High Risk'] = 1

# OR if Recommended Action includes serious follow-up
datav['High Risk'] = datav.apply(
    lambda row: 1 if 'BIOPSY' in row['Recommended Action'].upper() or 'COLPOSCOPY' in row['Recommended Action'].upper() else row['High Risk'],
    axis=1
)

# OR if Screening Type is suspicious
datav.loc[datav['Screening Type Suspicious'] == 1, 'High Risk'] = 1

# ✅ Done
print("High Risk class distribution:")
print(datav['High Risk'].value_counts())

print(datav['Screening Type Suspicious'].value_counts())

datav


In [ ]:
#one hot encoding.
#Here we encode the Region,screening Type Last to binary. 

df = pd.DataFrame({
    'Screening Type Last': ['Pap Smear', 'HPV DNA Test', 'Co-testing', 'Pap Smear'],
    'Region': ['Nairobi', 'Kisumu', 'Mombasa', 'Nairobi'],
    'Screening Type Last':['PAP SMEAR','HPV DNA'],
    'Recommended Action':
})

# Apply one-hot encoding
df_encoded = pd.get_dummies(df, columns=['Screening Type Last', 'Region'])
print(df_encoded)


In [ ]:
# drop some columns
#Region
#Reason to drop:
#This column is likely categorical geographic data (e.g., "Central", "Nairobi", "Unknown") and:
#It doesn’t directly indicate a person’s biological or clinical risk
#Might introduce location-based bias (e.g., model predicts higher risk just because someone is from a certain place)
#Can skew the model especially with small or uneven sample sizes per region

#Screening Type Last
#Reason to drop:
#You already extracted useful info from this column into a binary feature called:Screening Type Suspicious

#This feature captures whether the last screening method was highly diagnostic (e.g., HPV test, Pap smear, or Colposcopy). Keeping both would introduce redundancy.
#Drop to avoid duplication and multicollinearity

#Recommended Action
#Reason to drop:
#You already used this column to build your target variable High Risk, based on whether the action was “biopsy” or “colposcopy”.
#Keeping this column in the features would leak information that was already used to define the target — this is called data leakage in machine learning.
#Drop to prevent data leakage

#Insurance Covered
#Reason to drop:
#While important socially, it:
#Doesn’t represent actual clinical or biological risk
#Could lead the model to predict based on financial access rather than medical condition
#Introduces ethical concerns (e.g., model assumes uninsured = low risk, which is false)
#Drop for ethical, clinical, and fairness reasons

# List of columns to drop
cols_to_drop = [
    'Region',
    'Screening Type Last',
    'Recommended Action',
    'Insurance Covered'
]

# Drop them from the cleaned dataset
datav_clean = datav.drop(columns=cols_to_drop)

# Optional: confirm the updated columns
print("Remaining columns after drop:")
print(datav_clean.columns)
datav_clean


# STEP 4 Modeling Phase-Decision Tree Classifier

## Why Decision Tree?
We're using a Decision Tree model (CART) to classify patients as **High Risk** or **Not High Risk** based on medical and behavioral features.

---

## Advantages of Decision Trees
- **Low computational cost**: Trees are logarithmic in complexity, making them scalable for large datasets.
- **White-box model**: Easy to interpret and visualize. We can trace how decisions are made.
- **Minimal data preparation**: No need to normalize, scale, or encode dummy variables for numerical trees.

---

## Disadvantages of Decision Trees
- **Overfitting**: Trees may perfectly fit training data but perform poorly on new data. (We can use pruning or switch to Random Forests.)
- **Instability**: Small data changes can lead to completely different trees.
- **Class ambiguity**: Trees struggle if many examples have identical features but different labels (noisy/conflicting data).

---

##  Data Check Before Modeling
To verify if identical inputs have different outputs (inconsistent cases), we can run this loop:



In [ ]:
# Train & Evaluate a Decision Tree Classifier 1. Splitting the data

from sklearn.model_selection import train_test_split

# Define features and target
features = [
    'Age', 
    'Sexual Partners', 
    'First Sexual Activity Age',
    'Smoking Status',
    'STDs History',
    'Screening Type Suspicious'
]

X = datav_clean[features]
y = datav_clean['High Risk']

# Train-test split
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)


In [ ]:
#Train, Evaluate, and Interpret a Decision Tree Model
# 1. Train the Model
#Now that you've split your data into training and testing sets, you're ready to train a DecisionTreeClassifier.

from sklearn.tree import DecisionTreeClassifier

# Initialize the model
dtree = DecisionTreeClassifier(random_state=42)

# Train (fit) the model on training data
dtree.fit(X_train, y_train)


In [ ]:
#Make Predictions
# Once trained, predict on the test data and evaluate performance.
# Predict on the test set
y_pred = dtree.predict(X_test)


## Model from ScikitLearn


*   Uses CART algorithm, meaning that each leaf can only have two children. aka binary trees
*   ID3 algorithm could produce nodes with more than 2 children




In [ ]:
#Evaluate the Model
#We'll use accuracy, confusion matrix, and classification report to assess how well your model performs.

from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

# Accuracy
print("Accuracy:", accuracy_score(y_test, y_pred))

# Confusion Matrix
print("\n Confusion Matrix:")
print(confusion_matrix(y_test, y_pred))

# Precision, Recall, F1-score
print("\n Classification Report:")
print(classification_report(y_test, y_pred, target_names=["Not High Risk", "High Risk"]))



In [ ]:
# Hyperparameter Tuning (Grid Search)
#  Train using GridSearchCV
from sklearn.model_selection import GridSearchCV
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import make_scorer, f1_score

# Define model and grid
param_grid = {
    'max_depth': [4, 5, 6, 7, 8, 9, 10, 11],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4],
    'criterion': ['gini', 'entropy']
}

grid_search = GridSearchCV(
    estimator=DecisionTreeClassifier(random_state=42),
    param_grid=param_grid,
    scoring=make_scorer(f1_score),
    cv=5,
    n_jobs=-1,
    verbose=1
)

grid_search.fit(X_train, y_train)

# Extract best model
best_model = grid_search.best_estimator_
y_pred = best_model.predict(X_test)


In [ ]:
#Evaluate the Best Model, this is after the grid search and prediction:
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

print("Accuracy:", accuracy_score(y_test, y_pred))
print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred))
print("\nClassification Report:")
print(classification_report(y_test, y_pred, target_names=["Not High Risk", "High Risk"]))


In [ ]:
# Visualize the Tree 
# This helps you understand the decision logic your model has learned.

from sklearn.tree import plot_tree
import matplotlib.pyplot as plt

plt.figure(figsize=(15, 10))
plot_tree(
    best_model,
    feature_names=features,
    class_names=['Not High Risk', 'High Risk'],
    filled=True,
    rounded=True
)
plt.title("First Decision Tree")
plt.show()




In [ ]:
# Feature Importance
# See which features the tree considered most useful in classification.

import pandas as pd
import matplotlib.pyplot as plt

# Use best_model, not dtree
feature_importances = pd.Series(best_model.feature_importances_, index=features)

# Sort and plot
feature_importances.sort_values(ascending=True).plot(
    kind='barh',
    title='Feature Importances',
    figsize=(8, 5),
    color='skyblue'
)

plt.xlabel("Importance Score")
plt.tight_layout()
plt.show()




In [ ]:
#Pruning (To Prevent Overfitting)
#(to do limit tree depth, minimum samples, etc.)

# Create a pruned version of the tree
dtree_pruned = DecisionTreeClassifier(
    max_depth=3,               # limit depth to 3 levels (simpler tree)
    min_samples_split=10,      # at least 10 samples to consider a split
    random_state=42            # reproducibility
)

dtree_pruned.fit(X_train, y_train)  # Train on training set



In [ ]:
# Predict with the pruned tree
y_pred_pruned = dtree_pruned.predict(X_test)


In [ ]:
# Evaluate the pruned model
print("Pruned Tree Accuracy:", accuracy_score(y_test, y_pred_pruned))


In [ ]:
from sklearn.metrics import classification_report, confusion_matrix

print(confusion_matrix(y_test, y_pred_pruned))
print(classification_report(y_test, y_pred_pruned))


# Compare: Grid Search Model vs Pruned Tree
* Evaluate both models on the same test data
* Show accuracy, F1, confusion matrix, and classification report
* Output the better final decision tree

In [ ]:
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score, f1_score

# -------------------------------
# Evaluate Best Grid Search Model
# -------------------------------
y_pred_best = best_model.predict(X_test)

print("Grid Search Model Evaluation")
print("Accuracy:", accuracy_score(y_test, y_pred_best))
print("F1 Score:", f1_score(y_test, y_pred_best))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred_best))
print("Classification Report:\n", classification_report(y_test, y_pred_best))


# ---------------------
# Evaluate Pruned Model
# ---------------------
y_pred_pruned = dtree_pruned.predict(X_test)

print("\n Pruned Model Evaluation")
print("Accuracy:", accuracy_score(y_test, y_pred_pruned))
print("F1 Score:", f1_score(y_test, y_pred_pruned))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred_pruned))
print("Classification Report:\n", classification_report(y_test, y_pred_pruned))


# Summary of Results

| **Metric**           | **Grid Search Model**           | **Pruned Decision Tree**       |
|----------------------|----------------------------------|---------------------------------|
| **Accuracy**         | 0.95                             |    1.00                         |
| **F1 Score**         | 0.9677                           |    1.00                         |
| **Confusion Matrix** | 1 FN (missed class 0)            |    Perfect                      |
| **Interpretability** | Moderate (depth may vary)        |    Simple (you control depth)   |
| **Overfitting Risk** | Medium                           |    High (check generalizability) |

---

# Interpretation

##  Why the Pruned Model is Better (for Now):

* **Perfect performance on test set**: 100% accuracy and F1 score — no false positives or negatives.
* **Simpler model**: Easier to interpret and explain.
* **Faster**: Less computational load.

---

# Disadvantage: Overfitting Risk

Even though the pruned tree scores perfectly, keep in mind:

* It may be **memorizing the training data**.
* With a small test set (only 20 samples), this result may **not generalize** well to unseen patients.

---

# Final Steps

**stick with the pruned tree**, because:

* The **dataset is small**.
* We need a **quick, interpretable model**.
* Need a proof of concept**explain it to doctors or healthcare decision-makers**.

---

We compared a hyperparameter-tuned Decision Tree via Grid Search with a manually pruned version. Although both performed well, the pruned tree achieved perfect performance on our test set with fewer parameters and more interpretability. Due to our small dataset and the importance of explainability in medical predictions, we select the pruned model as our final classifier.



## Final decision tree model

In [ ]:
# Visualize the Final Tree-pruned tree
# imports are included first
from sklearn.tree import export_graphviz
import graphviz

#  Visualize the pruned decision tree
dot_data = export_graphviz(
    dtree_pruned,  # Make sure this is the correct pruned model variable
    out_file=None,
    feature_names=features,
    class_names=["Not High Risk", "High Risk"],
    filled=True,
    rounded=True,
    special_characters=True
)

graph = graphviz.Source(dot_data)
graph.render("final_decision_tree_pruned", format='png', cleanup=True)
graph  # This will display the tree in Jupyter Notebook



In [ ]:
# Save Your Model (Optional for deployment)
import joblib

# Save the trained model
joblib.dump(dtree_pruned, 'final_cervical_risk_model.pkl')


# Conclusion

After building, tuning, and evaluating our Decision Tree Classifier, here is the summary of our results:

---

## Optimal Model Parameters (from Grid Search)
- **Criterion**: `entropy`
- **Max Depth**: `4`
- **Min Samples Leaf**: `2`
- **Min Samples Split**: `6`

These parameters were found using GridSearchCV and 5-fold cross-validation with F1 score as the evaluation metric.

---

## Final Model Performance
- **Accuracy**: ~96%
- **F1 Score**: ~0.97
- **Confusion Matrix**: Very few false positives/negatives
- **Interpretability**: High, due to controlled depth and pruned tree structure

---

##  Why This Model Works Well
- Balances accuracy and generalizability.
- Performs well on both training and test data.
- Easy to interpret for healthcare professionals and stakeholders.
- Ideal for small to medium medical datasets.

---

## Final Notes
- While performance is excellent on this dataset, further evaluation on **larger or external datasets** is advised before clinical use.
- Always ensure models are validated across populations and reviewed with domain experts before deployment.

---

 **Next Steps for more interactive interface:**
- Deploy the model for interactive predictions.
- Integrate with a simple web app or API.
- Explore ensemble methods (Random Forest, XGBoost) for more robustness.
